experimental/list_formatter/src/list_formatter.rs

```diff
@@ -46,86 +48,138 @@ impl ListFormatter {
         Ok(Self { data, width })
     }
 
-    fn format_internal<'c, B>(
+    fn format_internal<'c, W: Writeable, S: ?Sized>(
         &'c self,
-        values: &[&str],
-        mut builder: B,
-        append_value: fn(B, &str) -> B,
-        append_literal: fn(B, &str) -> B,
-    ) -> B {
+        values: &[W],
+        sink: &mut S,
+        write_w_value: fn(&mut S, &W) -> fmt::Result,
+        write_m_value: fn(&mut S, &str) -> fmt::Result,
```

Comment:
Suggestion: Another way to architect this function is to coerce the format inputs to `FormattedStringLike`, and then you don't need to go back and forth with the closures.

Reply:
But then I'd have to define even more methods (`append_value` and `append_literal`) on the public `FormattedStringLike` trait...